service/kernel/select_indexed.cpp

```diff
@@ -69,61 +62,262 @@ void run_quick_select_simd( ExecutionContextIface& context,
     range.global(globalRange, status); DAAL_CHECK_STATUS_PTR(status);
     range.local(localRange, status);   DAAL_CHECK_STATUS_PTR(status);
 
-    KernelArguments args(9);
+    KernelArguments args(10);
     args.set(0, dataVectors, AccessModeIds::read);
     args.set(1, indexVectors, AccessModeIds::read);
     args.set(2, result.values, AccessModeIds::write);
     args.set(3, result.indices, AccessModeIds::write);
     args.set(4, rndSeq, AccessModeIds::read);
     args.set(5, nRndSeq);
     args.set(6, vectorSize);
-    args.set(7, K);
-    args.set(8, vectorOffset);
+    args.set(7, lastVectorSize);
+    args.set(8, K);
+    args.set(9, vectorOffset);
 
-    context.run(range, sum_kernel, args, status);
+    context.run(range, func_kernel, args, status);
 }
 
-QuickSelectIndexed::Result QuickSelectIndexed::select(const UniversalBuffer& dataVectors, const UniversalBuffer& indexVectors,
-                        const UniversalBuffer& rndSeq, uint32_t nRndSeq,
-                        uint32_t K, uint32_t nVectors, uint32_t vectorSize,
+void run_direct_select_simd( ExecutionContextIface& context,
+                            ClKernelFactoryIface& kernelFactory,
+                            const UniversalBuffer& dataVectors,
+                            uint32_t K,
+                            uint32_t nVectors,
+                            uint32_t vectorSize,
+                            uint32_t lastVectorSize,
+                            uint32_t vectorOffset,
+                            QuickSelectIndexed::Result& result,
+                            services::Status* status)
+{
+    auto func_kernel = kernelFactory.getKernel("direct_select_group");
+
+    const uint32_t maxWorkItemsPerGroup = 16;
+    KernelRange localRange(1, maxWorkItemsPerGroup);
+    KernelRange globalRange(nVectors, maxWorkItemsPerGroup);
+
+    KernelNDRange range(2);
+    range.global(globalRange, status); DAAL_CHECK_STATUS_PTR(status);
+    range.local(localRange, status);   DAAL_CHECK_STATUS_PTR(status);
+
+    KernelArguments args(6);
+    args.set(0, dataVectors, AccessModeIds::read);
+    args.set(1, result.values, AccessModeIds::write);
+    args.set(2, result.indices, AccessModeIds::write);
+    args.set(3, vectorSize);
+    args.set(4, lastVectorSize);
+    args.set(5, vectorOffset);
+
+    context.run(range, func_kernel, args, status);
+}
+
+void SelectIndexed::convert(const UniversalBuffer& indices, const UniversalBuffer& labels,
+                   uint32_t nVectors, uint32_t vectorSize, uint32_t vectorOffset, services::Status* status)
+{
+    auto index2labels = labels.template get<int>().toHost(ReadWriteMode::readOnly, status);
```

Comment:
using signed int as indexes is enough?

Reply:
Yes, and signed int is recommended for OpenCL kernels